src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -244,26 +264,81 @@ public boolean equals(Object other) {
     //=========== Undo-Redo =============================================================
     @Override
     public boolean hasHistory() {
-        return trackedAddressBook.hasHistory();
+        return currentStatePointer > 0;
     }
 
     @Override
     public boolean canRedo() {
-        return trackedAddressBook.canRedo();
+        return currentStatePointer < modelManagerStateList.size() - 1;
     }
 
     @Override
-    public void undoAddressBook() {
-        trackedAddressBook.undo();
+    public void redo() {
+        if (!canRedo()) {
+            throw new ModelManager.NoRedoableStateException();
+        }
+        currentStatePointer++;
+        resetData(modelManagerStateList.get(currentStatePointer));
     }
 
+    /**
+     * Saves a copy of the current {@code ModelManager} state at the end of the state list.
+     * Undone states are removed from the state list.
+     */
     @Override
-    public void commitAddressBook() {
-        trackedAddressBook.commit();
+    public void commit() {
+        removeStatesAfterCurrentPointer();
+        Person personCopy =
+                this.selectedPerson == null
+                ? null
+                : this.selectedPerson.getCopy();
+        //making a copy of current filteredPersons
+        UniquePersonList obListCopy = new UniquePersonList();
+        obListCopy.setPersons(this.addressBook.getPersonList());
+        FilteredList<Person> filteredListCopy = new FilteredList<Person>(obListCopy.asUnmodifiableObservableList());
+        filteredListCopy.setPredicate(this.filteredPersons.getPredicate());
+        //maybe causing undone states to malfunction
+
+        modelManagerStateList.add(new ReadOnlyModelManager(
+                new AddressBook(this.addressBook),
+                filteredListCopy, //gets edited when new command wrong
+                this.userPrefs.getCopy(),
+                personCopy,
+                this.themeProperty
```

Comment:
You forgot to create a copy for the `themeProperty` here, which would also cause problems with undoing/redoing the `theme` command.